src/librustc_middle/ty/context.rs

```diff
@@ -1181,6 +1179,32 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
+    /// Constructs a `TyKind::Error` type and registers a `delay_span_bug` to ensure it gets used.
+    #[track_caller]
+    pub fn err(self) -> Ty<'tcx> {
+        self.sess.delay_span_bug(
+            DUMMY_SP,
+            &format!(
+                "TyKind::Error constructed but no error reported. {}",
+                std::panic::Location::caller()
```

Comment:
What if we add `#[track_caller]` to `delay_span_bug` as well and have *it* pull the location out? Since that would benefit every single `delay_span_bug` call, making it more like the `span_bug!` macro (modulo formatting).

Also it might be cool to use the `TyCxAt` helper to have `tcx.at(span).err()` use the span for the `delay_span_bug`, since I assume many of the callers of `tcx.err()` have a span around.

Reply:
Perhaps I could open an issue for this? I'm not sure that I will have time to work on this a lot more.

Reply:
Which part? Slapping `#[track_caller]` onto functions "just works". The span part I guess it's trickier and you can open an issue about.

Reply:
Opened #72619